android/expoview/src/main/java/host/exp/exponent/RNObject.kt

```diff
@@ -0,0 +1,314 @@
+    }
+    return this
+  }
+
+  fun assign(obj: Any?) {
+    if (obj != null) {
+      clazz = obj.javaClass
+    }
+    instance = obj
+  }
+
+  fun get(): Any? {
+    return instance
+  }
+
+  fun rnClass(): Class<*>? {
+    return clazz
+  }
+
+  fun version(): String {
+    val versionedClassName = clazz!!.name
+    return if (versionedClassName.startsWith("abi")) {
+      val abiVersion = versionedClassName.split(".").toTypedArray()[0]
+      abiVersion.substring(3)
+    } else {
+      UNVERSIONED
+    }
+  }
+
+  fun construct(vararg args: Any?): RNObject {
+    try {
+      instance = getConstructorWithTypes(clazz, *objectsToClasses(*args)).newInstance(*args)
+    } catch (e: NoSuchMethodException) {
+      EXL.e(TAG, e)
+    } catch (e: InvocationTargetException) {
+      EXL.e(TAG, e)
+    } catch (e: InstantiationException) {
+      EXL.e(TAG, e)
+    } catch (e: IllegalAccessException) {
+      EXL.e(TAG, e)
+    }
+    return this
+  }
+
+  fun call(name: String, vararg args: Any?): Any? {
+    return callWithReceiver(instance, name, *args)
+  }
+
+  fun callRecursive(name: String, vararg args: Any?): RNObject? {
+    val result = call(name, *args) ?: return null
+    return wrap(result)
+  }
+
+  fun callStatic(name: String, vararg args: Any?): Any? {
+    return callWithReceiver(null, name, *args)
+  }
+
+  fun callStaticRecursive(name: String, vararg args: Any?): RNObject {
+    return wrap(callStatic(name, *args))
+  }
+
+  fun setField(name: String, value: Any) {
+    setFieldWithReceiver(instance, name, value)
+  }
+
+  fun setStaticField(name: String, value: Any) {
+    setFieldWithReceiver(null, name, value)
+  }
+
+  private fun callWithReceiver(receiver: Any?, name: String, vararg args: Any?): Any? {
+    try {
+      return getMethodWithTypes(clazz, name, *objectsToClasses(*args)).invoke(receiver, *args)
+    } catch (e: IllegalAccessException) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: InvocationTargetException) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: NoSuchMethodException) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: NoSuchMethodError) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: Throwable) {
+      EXL.e(TAG, "Runtime exception in RNObject when calling method $name: $e")
+    }
+    return null
+  }
+
+  private fun setFieldWithReceiver(receiver: Any?, name: String, value: Any) {
+    try {
+      getFieldWithType(clazz, name, value.javaClass)[receiver] = value
+    } catch (e: IllegalAccessException) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: NoSuchFieldException) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: NoSuchMethodError) {
+      EXL.e(TAG, e)
+      e.printStackTrace()
+    } catch (e: Throwable) {
+      EXL.e(TAG, "Runtime exception in RNObject when setting field $name: $e")
+    }
+  }
+
+  fun onHostResume(one: Any?, two: Any?) {
+    call("onHostResume", one, two)
```

Comment:
later might want to revisit whether these args are needed

Reply:
https://linear.app/expo/issue/ENG-1830/revisit-necessity-of-rnobjectonhostresume-args